File: cpp_app/can_protocol.cpp
#include "can_protocol.h"
#include <algorithm>
#include <stdexcept>
#include <cmath>

std::string chargerControlName(ChargerControl ctrl) {
    switch (ctrl) {
        case ChargerControl::START_CHARGING:    return "START_CHARGING";
        case ChargerControl::STOP_OUTPUTTING:   return "STOP_OUTPUTTING";
        case ChargerControl::HEATING_DC_SUPPLY: return "HEATING_DC_SUPPLY";
    }
    return "UNKNOWN";
}

// --- StatusFlags ---

StatusFlags StatusFlags::fromByte(uint8_t b) {
    StatusFlags f;
    f.hardware_failure      = (b & 0x01) != 0;
    f.over_temperature      = (b & 0x02) != 0;
    f.input_voltage_error   = (b & 0x04) != 0;
    f.starting_state        = (b & 0x08) != 0;
    f.communication_timeout = (b & 0x10) != 0;
    return f;
}

uint8_t StatusFlags::toByte() const {
    uint8_t b = 0;
    if (hardware_failure)      b |= 0x01;
    if (over_temperature)      b |= 0x02;
    if (input_voltage_error)   b |= 0x04;
    if (starting_state)        b |= 0x08;
    if (communication_timeout) b |= 0x10;
    return b;
}

bool StatusFlags::anyFault() const {
    return hardware_failure || over_temperature
        || input_voltage_error || communication_timeout;
}

// --- Message1 ---

std::array<uint8_t, 8> Message1::encode() const {
    int v_raw = std::clamp(static_cast<int>(std::round(voltage_setpoint * 10.0)), 0, 0xFFFF);
    int i_raw = std::clamp(static_cast<int>(std::round(current_setpoint * 10.0)), 0, 0xFFFF);

    std::array<uint8_t, 8> buf{};
    // Big-endian: voltage (2 bytes), current (2 bytes), control (1 byte), 3 padding
    buf[0] = static_cast<uint8_t>((v_raw >> 8) & 0xFF);
    buf[1] = static_cast<uint8_t>(v_raw & 0xFF);
    buf[2] = static_cast<uint8_t>((i_raw >> 8) & 0xFF);
    buf[3] = static_cast<uint8_t>(i_raw & 0xFF);
    buf[4] = static_cast<uint8_t>(control);
    buf[5] = 0;
    buf[6] = 0;
    buf[7] = 0;
    return buf;
}

Message1 Message1::decode(const uint8_t* data, size_t len) {
    if (len < 5)
        throw std::runtime_error("Message1 requires >= 5 bytes");

    uint16_t v_raw = (static_cast<uint16_t>(data[0]) << 8) | data[1];
    uint16_t i_raw = (static_cast<uint16_t>(data[2]) << 8) | data[3];
    uint8_t ctrl = data[4];

    if (ctrl > 2)
        throw std::runtime_error("Unknown ChargerControl value: " + std::to_string(ctrl));

    Message1 msg;
    msg.voltage_setpoint = v_raw / 10.0;
    msg.current_setpoint = i_raw / 10.0;
    msg.control = static_cast<ChargerControl>(ctrl);
    return msg;
}

// --- Message2 ---

std::array<uint8_t, 8> Message2::encode() const {
    int v_raw   = std::clamp(static_cast<int>(std::round(output_voltage * 10.0)), 0, 0xFFFF);
    int i_raw   = std::clamp(static_cast<int>(std::round(output_current * 10.0)), 0, 0xFFFF);
    int vin_raw = std::clamp(static_cast<int>(std::round(input_voltage * 10.0)), 0, 0xFFFF);
    int tmp_raw = std::clamp(static_cast<int>(std::round(temperature + 40.0)), 0, 0xFF);

    std::array<uint8_t, 8> buf{};
    // Big-endian: output_voltage(2), output_current(2), status(1), input_voltage(2), temperature(1)
    buf[0] = static_cast<uint8_t>((v_raw >> 8) & 0xFF);
    buf[1] = static_cast<uint8_t>(v_raw & 0xFF);
    buf[2] = static_cast<uint8_t>((i_raw >> 8) & 0xFF);
    buf[3] = static_cast<uint8_t>(i_raw & 0xFF);
    buf[4] = status.toByte();
    buf[5] = static_cast<uint8_t>((vin_raw >> 8) & 0xFF);
    buf[6] = static_cast<uint8_t>(vin_raw & 0xFF);
    buf[7] = static_cast<uint8_t>(tmp_raw);
    return buf;
}

Message2 Message2::decode(const uint8_t* data, size_t len) {
    if (len < 8)
        throw std::runtime_error("Message2 requires 8 bytes");

    uint16_t v_raw   = (static_cast<uint16_t>(data[0]) << 8) | data[1];
    uint16_t i_raw   = (static_cast<uint16_t>(data[2]) << 8) | data[3];
    uint8_t  st_byte = data[4];
    uint16_t vin_raw = (static_cast<uint16_t>(data[5]) << 8) | data[6];
    uint8_t  tmp_raw = data[7];

    Message2 msg;
    msg.output_voltage = v_raw / 10.0;
    msg.output_current = i_raw / 10.0;
    msg.status = StatusFlags::fromByte(st_byte);
    msg.input_voltage = vin_raw / 10.0;
    msg.temperature = static_cast<double>(tmp_raw) - 40.0;
    return msg;
}

File: obc_controller/can_protocol.py
"""
CAN protocol codec for OBC charger communication.

Based on CAN BUS COMMUNICATION SPECIFICATION v1.3:
  - J1939, CAN2.0B, 29-bit extended frame
  - Default bus speed: 250 kbps
  - Message1: BMS -> OBC  (ID 0x1806E5F4, 500 ms)
  - Message2: OBC -> BCA  (ID 0x18FF50E5, 500 ms)
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Optional


# ---------------------------------------------------------------------------
# Node source addresses (J1939)
# ---------------------------------------------------------------------------
SA_BMS = 0xF4   # 244
SA_OBC = 0xE5   # 229
SA_BCA = 0x50   # 80

# ---------------------------------------------------------------------------
# CAN message IDs (29-bit extended)
# ---------------------------------------------------------------------------
MSG1_ID = 0x1806E5F4   # BMS -> OBC
MSG2_ID = 0x18FF50E5   # OBC -> BCA

# ---------------------------------------------------------------------------
# Timing
# ---------------------------------------------------------------------------
CYCLE_MS = 500          # Transmit period (ms)
TIMEOUT_S = 5.0         # Communication timeout (s)


class ChargerControl(IntEnum):
    START_CHARGING = 0
    STOP_OUTPUTTING = 1
    HEATING_DC_SUPPLY = 2


# ---------------------------------------------------------------------------
# Message1  (BMS -> OBC)
# ---------------------------------------------------------------------------
@dataclass
class Message1:
    """Command message sent from BMS (this app) to OBC."""
    voltage_setpoint: float = 0.0   # V, resolution 0.1 V
    current_setpoint: float = 0.0   # A, resolution 0.1 A
    control: ChargerControl = ChargerControl.STOP_OUTPUTTING

    def encode(self) -> bytes:
        """Encode to 8-byte CAN payload (big-endian, unsigned)."""
        v_raw = int(round(self.voltage_setpoint * 10))
        i_raw = int(round(self.current_setpoint * 10))
        v_raw = max(0, min(v_raw, 0xFFFF))
        i_raw = max(0, min(i_raw, 0xFFFF))
        return struct.pack(">HHB3x", v_raw, i_raw, int(self.control))

    @classmethod
    def decode(cls, data: bytes) -> "Message1":
        """Decode 8-byte CAN payload."""
        if len(data) < 5:
            raise ValueError(f"Message1 requires >= 5 bytes, got {len(data)}")
        v_raw, i_raw, ctrl = struct.unpack_from(">HHB", data, 0)
        try:
            control = ChargerControl(ctrl)
        except ValueError:
            raise ValueError(
                f"Unknown ChargerControl value: {ctrl} "
                f"(expected 0, 1, or 2)"
            )
        return cls(
            voltage_setpoint=v_raw / 10.0,
            current_setpoint=i_raw / 10.0,
            control=control,
        )


# ---------------------------------------------------------------------------
# Status flags in Message2 BYTE4
# ---------------------------------------------------------------------------
@dataclass
class StatusFlags:
    hardware_failure: bool = False       # bit0: 0=Normal, 1=Failure
    over_temperature: bool = False       # bit1: 0=Normal, 1=Over-temp
    input_voltage_error: bool = False    # bit2: 0=Normal, 1=Wrong
    starting_state: bool = False         # bit3: 0=Charging, 1=Stays off
    communication_timeout: bool = False  # bit4: 0=Normal, 1=Timeout

    @classmethod
    def from_byte(cls, b: int) -> "StatusFlags":
        return cls(
            hardware_failure=bool(b & 0x01),
            over_temperature=bool(b & 0x02),
            input_voltage_error=bool(b & 0x04),
            starting_state=bool(b & 0x08),
            communication_timeout=bool(b & 0x10),
        )

    def to_byte(self) -> int:
        b = 0
        if self.hardware_failure:
            b |= 0x01
        if self.over_temperature:
            b |= 0x02
        if self.input_voltage_error:
            b |= 0x04
        if self.starting_state:
            b |= 0x08
        if self.communication_timeout:
            b |= 0x10
        return b

    def any_fault(self) -> bool:
        return (
            self.hardware_failure
            or self.over_temperature
            or self.input_voltage_error
            or self.communication_timeout
        )


# ---------------------------------------------------------------------------
# Message2  (OBC -> BCA)
# ---------------------------------------------------------------------------
@dataclass
class Message2:
    """Telemetry message broadcast by OBC."""
    output_voltage: float = 0.0    # V
    output_current: float = 0.0    # A
    status: StatusFlags = field(default_factory=StatusFlags)
    input_voltage: float = 0.0     # V
    temperature: float = 0.0       # degC

    def encode(self) -> bytes:
        """Encode to 8-byte CAN payload."""
        v_raw = int(round(self.output_voltage * 10))
        i_raw = int(round(self.output_current * 10))
        vin_raw = int(round(self.input_voltage * 10))
        temp_raw = int(round(self.temperature + 40))
        v_raw = max(0, min(v_raw, 0xFFFF))
        i_raw = max(0, min(i_raw, 0xFFFF))
        vin_raw = max(0, min(vin_raw, 0xFFFF))
        temp_raw = max(0, min(temp_raw, 0xFF))
        return struct.pack(
            ">HHBHB",
            v_raw, i_raw, self.status.to_byte(), vin_raw, temp_raw,
        )

    @classmethod
    def decode(cls, data: bytes) -> "Message2":
        """Decode 8-byte CAN payload."""
        if len(data) < 8:
            raise ValueError(f"Message2 requires 8 bytes, got {len(data)}")
        v_raw, i_raw, status_byte, vin_raw, temp_raw = struct.unpack_from(
            ">HHBHB", data, 0
        )
        return cls(
            output_voltage=v_raw / 10.0,
            output_current=i_raw / 10.0,
            status=StatusFlags.from_byte(status_byte),
            input_voltage=vin_raw / 10.0,
            temperature=temp_raw - 40,
        )
